### PR TITLE
Various improvements (code duplication, pathlib)

### LINKED ISSUE #50: Use pathlib instead of os
See reviews on 

https://github.com/SSH-key-rotation-AWS/team-henrique/pull/46
https://github.com/SSH-key-rotation-AWS/team-henrique/pull/47

### LINKED ISSUE #52: Improve security for key generation
See https://github.com/SSH-key-rotation-AWS/team-henrique/pull/46#discussion_r1247893548

### LINKED ISSUE #69: Localize key paths to ONE spot - avoid repetition
The path is currently spelled out in a bunch of places - fix that.

### PR DESCRIPTION
**Pull request by ygins**
- use pathlib
- create directories/files with pathlib - permissions from otuset improves security
- publisher calls datastore
- various bugfixes w/ regards to directories not deleting
- full implementation of temp directories for both local & S3 publisher

closes #50 
closes #69 
closes #52

### CODE REVIEW

File: switcheroo/custom_keygen.py [hunk 1]
```diff
@@ -105,37 +104,37 @@ def generate_private_public_key(cls) -> tuple[bytes, bytes]:
...
         return private_key, public_key
 
+    @classmethod
+    def store_private_key(cls, private_key: bytes, private_key_dir: Path):
+        private_key_dir.mkdir(parents=True, exist_ok=True)
+        private_key_path = private_key_dir / KeyGen.PRIVATE_KEY_NAME
+        # If old key is there, delete it
+        private_key_path.unlink(missing_ok=True)
+        private_key_path.touch(mode=0o600)
+        with open(private_key_path, mode="wb") as private_out:
```
Comment (thehcma) [inline switcheroo/custom_keygen.py:114]: we are still not using an `opener` - this is better, but you're still doing more work than needed

File: switcheroo/custom_keygen.py [hunk 2]
```diff
@@ -105,37 +104,37 @@ def generate_private_public_key(cls) -> tuple[bytes, bytes]:
         public_key = key.public_key().export_key(format="OpenSSH")
         return private_key, public_key
 
+    @classmethod
+    def store_private_key(cls, private_key: bytes, private_key_dir: Path):
+        private_key_dir.mkdir(parents=True, exist_ok=True)
+        private_key_path = private_key_dir / KeyGen.PRIVATE_KEY_NAME
+        # If old key is there, delete it
+        private_key_path.unlink(missing_ok=True)
```
Comment (thehcma) [inline switcheroo/custom_keygen.py:112]: why is this needed?

File: switcheroo/custom_keygen.py [hunk 3]
```diff
@@ -105,37 +104,37 @@ def generate_private_public_key(cls) -> tuple[bytes, bytes]:
...
+        private_key_path.touch(mode=0o600)
+        with open(private_key_path, mode="wb") as private_out:
+            private_out.write(private_key)
+
+    @classmethod
+    def store_public_key(cls, public_key: bytes, public_key_dir: Path):
+        public_key_dir.mkdir(parents=True, exist_ok=True)
+        public_key_path = public_key_dir / KeyGen.PUBLIC_KEY_NAME
+        # If old key is there, delete it
+        public_key_path.unlink(missing_ok=True)
```
Comment (thehcma) [inline switcheroo/custom_keygen.py:122]: why is this needed?

File: switcheroo/data_stores.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+from switcheroo import paths
+from switcheroo.custom_keygen import KeyGen, KeyMetadata
+
+
+class DataStore(ABC):
+    "A server uses a DataStore to get public keys from somewhere."
+
+    def __init__(self, ssh_home: Path | None = None, temp: bool = False):
+        # If temp is true, will reset the home_dir to a temp file upon usage as a context manager
+        self.temp = temp
```
Comment (thehcma) [inline switcheroo/data_stores.py:15]: private member should start with a leading _

File: switcheroo/data_stores.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        # ssh_home is only relevant if not being used as a context manager
+        self._dir: Path = paths.local_ssh_home() if ssh_home is None else ssh_home
+        # To be used later if we decide to use this as a context manager, and temp is selected
+        self._temp_dir: TemporaryDirectory[str] | None = None
```
Comment (thehcma) [inline switcheroo/data_stores.py:19]: not sure this makes sense - this class should be an abstract base class - what's the point of having this if the S3-based implementation will be using this

File: switcheroo/data_stores.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+    @property
+    def home_dir(self) -> Path:
```
Comment (thehcma) [inline switcheroo/data_stores.py:22]: what's the point of this in the S3 case?

File: switcheroo/data_stores.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+                prefix="ssh-keys-", dir=self._dir
+            )
+
+    def __exit__(self, exc_t, exc_v, exc_tb):
+        if self.temp:
+            assert self._temp_dir is not None
+            self._temp_dir.__exit__(None, None, None)
+
+
+class S3DataStore(DataStore):
```
Comment (thehcma) [inline switcheroo/data_stores.py:60]: this should be in a different module - otherwise to use the filesystem based implementation I have to drag in all of the `boto` stuff

File: switcheroo/data_stores.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    ):
+        super().__init__(ssh_home=ssh_home, temp=temp)
+        self._s3_bucket_name = _s3_bucket_name
+
+    @property
+    def s3_bucket_name(self):
+        "The name of the bucket in which the keys are stored"
+        return self._s3_bucket_name
+
+    def get_sshd_config_line(self) -> str:
```
Comment (thehcma) [inline switcheroo/data_stores.py:74]: this doesn't seem to belong here it has nothing to do with the data store

**Review by thehcma — APPROVED** [review]
I am approving, but the design of this is still a bit messy - we seem to be combining things that don't belong together

We need to work on having a better separation of concerns:

* the logic to store data
* the logic to create keys

File: switcheroo/custom_keygen.py (hunk 1 above)
Comment (ygins) [reply switcheroo/custom_keygen.py:114]: 👍

File: switcheroo/custom_keygen.py (hunk 2 above)
Comment (ygins) [reply switcheroo/custom_keygen.py:112]: 👍

File: switcheroo/custom_keygen.py (hunk 3 above)
Comment (ygins) [reply switcheroo/custom_keygen.py:122]: 👍

File: switcheroo/data_stores.py (hunk 4 above)
Comment (ygins) [reply switcheroo/data_stores.py:15]: 👍

File: switcheroo/data_stores.py (hunk 5 above)
Comment (ygins) [reply switcheroo/data_stores.py:19]: For private key storage - if we're testing the S3DataStore, having this logic also takes care of deleting private keys when we are done with them

File: switcheroo/data_stores.py (hunk 6 above)
Comment (ygins) [reply switcheroo/data_stores.py:22]: see above

File: switcheroo/data_stores.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    def publish_with_metadata(
+        self, host: str, user: str, metadata: KeyMetadata | None
+    ) -> tuple[str, KeyMetadata]:
+        """Publish a new public key for the given host and user with metadata"""
+
+    @abstractmethod
+    def retrieve(self, host: str, user: str) -> str:
+        """Retrieve the public key for the given host and user"""
+
+    def __enter__(self):
```
[the remark this answers is not in the document]
Comment (ygins) [reply switcheroo/data_stores.py:46]: I guess it's not necessary but it does specify common behavior in what a DataStore does - it makes the retrieve_keys.py script nicer as well

File: switcheroo/data_stores.py (hunk 7 above)
Comment (ygins) [reply switcheroo/data_stores.py:60]: 👍

File: switcheroo/data_stores.py (hunk 8 above)
Comment (ygins) [reply switcheroo/data_stores.py:74]: Are you referring to the sshd_config_line method? If so I see, we can extract that.

File: switcheroo/data_stores.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    @property
+    def s3_bucket_name(self):
+        "The name of the bucket in which the keys are stored"
+        return self._s3_bucket_name
+
+    def get_sshd_config_line(self) -> str:
+        return f"s3 {self.s3_bucket_name}"
+
+    def retrieve(self, host: str, user: str):
+        s3_client = boto3.client("s3")
```
[the remark this answers is not in the document]
Comment (ygins) [reply switcheroo/data_stores.py:78]: 👍

File: switcheroo/data_stores.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+        # Store the new public key in S3 bucket
+        s3_client = boto3.client("s3")
+        s3_client.put_object(
+            Body=public_key,
+            Bucket=self.s3_bucket_name,
+            Key=str(paths.cloud_public_key_loc(host, user)),
+        )
+
+        # Store the private key on the local machine
+        KeyGen.store_private_key(
```
[the remark this answers is not in the document]
Comment (ygins) [reply switcheroo/data_stores.py:97]: 👍

File: switcheroo/data_stores.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+            private_key=private_key,
+            private_key_dir=paths.local_key_dir(host, user, home_dir=self.home_dir),
+        )
+
+        return public_key.decode()
+
+    def publish_with_metadata(
```
[the remark this answers is not in the document]
Comment (ygins) [reply switcheroo/data_stores.py:104]: see above, definitely open to changing

File: switcheroo/data_stores.py (hunk 8 above)
Comment (ygins) [reply switcheroo/data_stores.py:74]: However, would it then just be a check in the Server class? ie the server checks which data store is being used and generates the command line accordingly? If so, then we lose abstraction

File: switcheroo/data_stores.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ygins) [reply switcheroo/data_stores.py:97]: While I agree it doesn't make so much sense for key writing/storing logic to be in a class called KeyGen, does it necessarily have to be in the Publisher method itself? It would be helpful to have it as a utility function called by the DataStore for a few reasons:

- The S3Publisher still writes private keys to files - it can just call that utility method (for the private key, at least)
- Unit testing the server - want to avoid using a Publisher. Better to call a utility method

Thanks!

**Comment by ygins** [comment]
I have some follow ups which I commented here. Merging now to move forwards with other features, and will discuss these & change what needs to be changed soon
